Firebase Push Notifications.md

```diff
@@ -0,0 +1,111 @@
+
+ * CONS
+    * User can get notification only at last logged in device.
+
+    * If user A login through his own phone, and then user A login and logout through user B phone(because user A left phone at home). All push notifications sent to user A in between will not be received because user A hasn't device_id in DB and user A has to login again through his phone.
+
+  * PROS
+    * Easy to build.
+
+2. Approach based on multiple device ids
+
+  * We need 1:N relationships between User and UserDevice models.
+
+  * Create same API endpoints for creating and deleting device_ids like in previous example.
+  * PROS
+    * User can get push notifications at multiple devices.
+  * CONS
+    * We need to maintain user device_ids(create, delete).
+
+3. Approach based on topics
+
+  * Topic is group of one or more device_ids saved in Firebase by specific name. Name can be anything that matches the regular expression: `[a-zA-Z0-9-_.~%]+`
+
+  * Imagine you are building app like Reddit.
+
+
+  * Persisted topics
+    * In this approach we have stored topic names in DB. For example in every subreddit record.
+    * We can have one topic for every subreddit.
+    * We can't create new topic from server side and because of that mobile client has to create topics. They can do this explicit after subreddit creation, or implicit after subscription. (Thats how Firebase works when topic doesn't exist. If topic doesn't exist in Firebase, it will be created and user will be subscribed after creation)
+    * At every login, mobile app fetches subscribed subreddits from our API and they subscribe user to topic/topics on Firebase.
+    * At every logout mobile app unsubscribe user(only current device) from topic/topics on Firebase.
+    * CONS
+      * We are duplicating topic names in our DB and Firebase.
+
+
+  * Non Persisted topics
+    * We can make convention for topic names with mobile developers.
+    * Every topic name is parameterized subreddit name.
+    (Firebase Push Notifications -> firebase-push-notifications)
+    * PROS:
+      * With this approach we don't need to store topic names in DB.
+      * We don't need topic_id field in DB because mobile team can get them through subscribed subreddits names.
+      * Easy to build.
+    * CONS:
+      * It doesn't seems that Firebase has topic limits but maybe in future they will restrict this like Amazon SNS.
+
+## Choosing right development approach
+When we are choosing between development approaches there are few questions for which we need answer. If we know answer at all questions, then we can choose development approach which is best for our application.
+
+1. How many devices user will be using?
+  * One device: Any approach
+  * Multiple devices: Second or third approach
+2. Are notifications grouped by any rules in our application or we always need to send notification to one specific user or to all users together?
```

Comment:
Here is also a problem of custom made topics. This topic needs to be made on serverside by an admin. This gets more complicated because you need mobile app to first connect to firebase to get's its instance-id

Please research this further.